src/dataset.rs

```diff
@@ -378,6 +446,83 @@ impl Drop for Dataset {
     }
 }
 
+/// This is a wrapper type for passing a thread-safe [`Dataset`] across threads.
+///
+/// Only read-only rasters can be thread-safe.
+///
+/// Note: most drivers don't support native thread-safety, LIBRETIFF being a
+/// notable exception. For the others, GDAL will reopen the dataset multiple
+/// times. On Unix systems, you might want to bump the open files limit (for
+/// example, see `ulimit` and the `rlimit` crate).
+///
+/// # Example
+/// ```no_run
+/// use gdal::{Dataset, DatasetOptions, GdalOpenFlags};
+/// use std::thread;
+///
+/// let ds = Dataset::open_ex(
+///     "file.tif",
+///     DatasetOptions {
+///         open_flags: GdalOpenFlags::GDAL_OF_RASTER | GdalOpenFlags::GDAL_OF_THREAD_SAFE,
+///         ..Default::default()
+///     },
+/// )?;
+/// let thread_safe_ds = ds.try_into_thread_safe(GdalOpenFlags::GDAL_OF_RASTER)?;
+///
+/// // Clone and use across threads
+/// thread::scope(|s| {
+///     for _ in 0..4 {
+///         let ds = thread_safe_ds.clone();
+///         s.spawn(move || {
+///             let band = ds.rasterband(1)?;
+///             // ... read from band ...
+///             Ok::<_, gdal::errors::GdalError>(())
+///         });
+///     }
+/// });
+/// # Ok::<_, gdal::errors::GdalError>(())
+/// ```
+#[cfg(any(all(major_ge_3, minor_ge_10), major_ge_4))]
+#[derive(Clone, Debug)]
+#[non_exhaustive]
+pub struct ThreadSafeDataset {
+    inner: Arc<Dataset>,
+}
+
+#[cfg(any(all(major_ge_3, minor_ge_10), major_ge_4))]
+unsafe impl Sync for ThreadSafeDataset {}
+#[cfg(any(all(major_ge_3, minor_ge_10), major_ge_4))]
+unsafe impl Send for ThreadSafeDataset {}
+
+#[cfg(any(all(major_ge_3, minor_ge_10), major_ge_4))]
+impl ThreadSafeDataset {
+    /// # Safety
+    /// Dataset must be thread-safe.
+    unsafe fn new(dataset: Dataset) -> Self {
+        Self {
+            // `GDALReferenceDataset` is not thread-safe, so we keep our own reference count using an `Arc`.
+            #[allow(clippy::arc_with_non_send_sync)]
+            inner: Arc::new(dataset),
+        }
+    }
+}
+
+#[cfg(any(all(major_ge_3, minor_ge_10), major_ge_4))]
+impl AsRef<Dataset> for ThreadSafeDataset {
+    fn as_ref(&self) -> &Dataset {
+        &self.inner
+    }
+}
+
+#[cfg(any(all(major_ge_3, minor_ge_10), major_ge_4))]
```

Comment:
For style purposes: Could we make a gated module and impl this in this module? Then, we would only have to import/re-export it once in `dataset.rs` and get rid of 90% of the version gates?

Reply:
Done.

Reply:
@ChristianBeilschmidt can you take another look? Thanks!